### PR TITLE
[SYCL] Implement more verbose error handling.

### PR DESCRIPTION
**Pull request by rbegam**
This ensures that no exception is silently ignored by sycl. This
also includes API names and error codes in the error messages.

Signed-off-by: rehana begam <rehana.begam@intel.com>

### CODE REVIEW

File: sycl/source/detail/device_info.hpp [hunk 1]
```diff
@@ -969,7 +969,11 @@ template <> struct get_device_info<bool, info::device::usm_device_allocations> {
         dev, pi::cast<RT::PiDeviceInfo>(info::device::usm_device_allocations),
         sizeof(pi_usm_capabilities), &caps, nullptr);
 
-    return (Err != PI_SUCCESS) ? false : (caps & PI_USM_ACCESS);
+    if (Err != PI_SUCCESS) {
+      Plugin.reportPiError(Err, "piDeviceGetInfo()");
```
Comment (smaslov-intel) [inline sycl/source/detail/device_info.hpp:973]: I think we need not throw the error to the user; just returning "false" for this query? 
Same comment in other queries where we can "deal" with the error.

**Review by smaslov-intel — CHANGES_REQUESTED** [review]

**Comment by smaslov-intel** [comment]
> includes API names and error codes in the error messages.

I am not sure it is a good idea to show PI API names to the user since they aren't documented anywhere.
Should we instead give the name of SYCL method where that fail occurs?

File: sycl/source/detail/device_info.hpp (hunk 1 above)
Comment (rbegam) [reply sycl/source/detail/device_info.hpp:973]: will restore the old code here.

File: sycl/source/detail/global_handler.cpp [hunk 2]
```diff
@@ -140,7 +140,12 @@ void shutdown() {
       // some parameters in the plugin tear-down process.
       // Currently, it is not used.
       void *PluginParameter = nullptr;
-      Plugin.call_nocheck<PiApiKind::piTearDown>(PluginParameter);
+      auto Error = Plugin.call_nocheck<PiApiKind::piTearDown>(PluginParameter);
+
+      if (Error != PI_SUCCESS) {
+        Plugin.reportPiError(Error, "piTearDown()");
+      }
```
Comment (s-kanaev) [inline sycl/source/detail/global_handler.cpp:147]: Why do we need `call_nocheck` here if we end up throwing an exception when smth different from `PI_SUCCESS` is returned? I think it's shorter to use `Plugin.call` here.

File: sycl/source/detail/pi.cpp [hunk 3]
```diff
@@ -141,8 +141,12 @@ void contextSetExtendedDeleter(const cl::sycl::context &context,
   auto impl = getSyclObjImpl(context);
   auto contextHandle = reinterpret_cast<pi_context>(impl->getHandleRef());
   auto plugin = impl->getPlugin();
-  plugin.call_nocheck<PiApiKind::piextContextSetExtendedDeleter>(
+  auto Error = plugin.call_nocheck<PiApiKind::piextContextSetExtendedDeleter>(
       contextHandle, func, user_data);
+
+  if (Error != PI_SUCCESS) {
+    plugin.reportPiError(Error, "piextContextSetExtendedDeleter()");
+  }
```
Comment (s-kanaev) [inline sycl/source/detail/pi.cpp:149]: Same question here.

File: sycl/source/detail/scheduler/commands.cpp [hunk 4]
```diff
@@ -1762,6 +1762,11 @@ pi_result ExecCGCommand::SetKernelParamsAndLaunch(
       MQueue->getHandleRef(), Kernel, NDRDesc.Dims, &NDRDesc.GlobalOffset[0],
       &NDRDesc.GlobalSize[0], LocalSize, RawEvents.size(),
       RawEvents.empty() ? nullptr : &RawEvents[0], &Event);
+
+  if (Error != PI_SUCCESS) {
+    Plugin.reportPiError(Error, "piEnqueueKernelLaunch()");
+  }
+
```
Comment (s-kanaev) [inline sycl/source/detail/scheduler/commands.cpp:1769]: Same question here.

File: sycl/include/CL/sycl/detail/common.hpp [hunk 5]
```diff
@@ -104,6 +104,16 @@ static inline std::string codeToString(cl_int code) {
   /* ":" __SYCL_STRINGIFY_LINE(__LINE__) ": " */                               \
                                "Native API returns: "
 
+#define __SYCL_REPORT_PLUGIN_ERR(expr, name, exc)                              \
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/detail/common.hpp:107]: Could you please clarify if we can convert it to a function?

File: sycl/source/detail/global_handler.cpp (hunk 2 above)
Comment (rbegam) [reply sycl/source/detail/global_handler.cpp:147]: good catch, got overlooked. I'll use `Plugin.call`

File: sycl/include/CL/sycl/detail/common.hpp (hunk 5 above)
Comment (rbegam) [reply sycl/include/CL/sycl/detail/common.hpp:107]: will convert it.

File: sycl/source/detail/global_handler.cpp (hunk 2 above)
Comment (rbegam) [reply sycl/source/detail/global_handler.cpp]: done.

File: sycl/source/detail/pi.cpp (hunk 3 above)
Comment (rbegam) [reply sycl/source/detail/pi.cpp]: done

File: sycl/include/CL/sycl/detail/common.hpp (hunk 5 above)
Comment (rbegam) [reply sycl/include/CL/sycl/detail/common.hpp]: done

File: sycl/source/detail/device_info.hpp (hunk 1 above)
Comment (rbegam) [reply sycl/source/detail/device_info.hpp]: done

File: sycl/source/detail/plugin.hpp [hunk 6]
```diff
@@ -56,6 +56,15 @@ class plugin {
     __SYCL_CHECK_OCL_CODE_THROW(pi_result, Exception);
   }
 
+  void reportPiError(RT::PiResult pi_result, string_class api_name) const {
```
Comment (smaslov-intel) [inline sycl/source/detail/plugin.hpp:59]:
```suggestion
  void reportPiError(RT::PiResult pi_result, string_class context) const {
```

Comment (smaslov-intel) [reply sycl/source/detail/plugin.hpp:59]: Add function comment

**Review by smaslov-intel — CHANGES_REQUESTED** [review]

File: sycl/source/detail/plugin.hpp (hunk 6 above)
Comment (romanovvlad) [reply sycl/source/detail/plugin.hpp:59]: Suggest also taking replacing `string_class` with `std::string &`

**Review by romanovvlad — COMMENTED** [review]
LGTM if comments with `reportPIError`  are fixed.

File: sycl/source/detail/device_info.hpp (hunk 1 above)
Comment (rbegam) [reply sycl/source/detail/device_info.hpp]: done.

File: sycl/source/detail/global_handler.cpp (hunk 2 above)
Comment (rbegam) [reply sycl/source/detail/global_handler.cpp]: done.

File: sycl/source/detail/plugin.hpp (hunk 6 above)
Comment (rbegam) [reply sycl/source/detail/plugin.hpp]: done.

**Review by smaslov-intel — DISMISSED** [review]

File: sycl/source/detail/program_impl.cpp [hunk 7]
```diff
@@ -362,7 +362,8 @@ void program_impl::create_cl_program_with_source(const std::string &Source) {
   }
 
   if (Err != PI_SUCCESS) {
-    Plugin.reportPiError(Err, "create_cl_program_with_source()");
+    std::string CallingFunction = "create_cl_program_with_source()";
+    Plugin.reportPiError(Err, CallingFunction);
```
Comment (smaslov-intel) [inline sycl/source/detail/program_impl.cpp:366]: what's wrong with prev version?

Comment (rbegam) [reply sycl/source/detail/program_impl.cpp:366]: there were compile errors as `std::string &` requires a `lvalue`.

Comment (smaslov-intel) [reply sycl/source/detail/program_impl.cpp:366]: Maybe use std::string_view?

**Comment by rbegam** [comment]
ping @smaslov-intel @romanovvlad

File: sycl/source/detail/program_impl.cpp (hunk 7 above)
Comment (smaslov-intel) [reply sycl/source/detail/program_impl.cpp:366]: it is kind of ugly that you cannot pass C-string directly, either use string_view or just "const char*"

**Review by smaslov-intel — CHANGES_REQUESTED** [review]

File: sycl/source/detail/program_impl.cpp (hunk 7 above)
Comment (rbegam) [reply sycl/source/detail/program_impl.cpp:366]: agree. how about using `std::string` instead of `std::string &`? that will resolve the lvalue issue.

Comment (smaslov-intel) [reply sycl/source/detail/program_impl.cpp:366]: This way there would be a copy made during the call. How about using "std::string &" but passing as string("bla-bla-bla")

Comment (rbegam) [reply sycl/source/detail/program_impl.cpp]: this didn't work either. i am using `const char*`

**Review by smaslov-intel — APPROVED** [review]

**Pull request merged**